server/services/v1/cache/query_runner.go

```diff
@@ -0,0 +1,312 @@
+	project, err := tenant.GetDatabase(ctx, metadata.NewDatabaseNameWithBranch(runner.req.GetProject(), metadata.MainBranch))
+	if err != nil {
+		return nil, err
+	}
+	encodedCacheTableName, err := runner.encoder.EncodeCacheTableName(tenant.GetNamespace().Id(), project.Id(), runner.req.GetName())
+	if err != nil {
+		return nil, err
+	}
+
+	options := &cache.GetOptions{
+		Expiry:    0,
+		GetDelete: false,
+	}
+	data, err := runner.cacheStore.Get(ctx, encodedCacheTableName, runner.req.GetKey(), options)
+	if err != nil {
+		if err == cache.ErrKeyNotFound {
+			return nil, errors.NotFound(err.Error())
+		}
+		return nil, err
+	}
+	return &CacheResponse{
+		Data: data.GetRawData(),
+	}, nil
+}
+
+func (runner *DelQueryRunner) Run(ctx context.Context, tenant *metadata.Tenant) (*CacheResponse, error) {
+	project, err := tenant.GetDatabase(ctx, metadata.NewDatabaseNameWithBranch(runner.req.GetProject(), metadata.MainBranch))
+	if err != nil {
+		return nil, err
+	}
+	encodedCacheTableName, err := runner.encoder.EncodeCacheTableName(tenant.GetNamespace().Id(), project.Id(), runner.req.GetName())
+	if err != nil {
+		return nil, err
+	}
+
+	deletedCount, err := runner.cacheStore.Delete(ctx, encodedCacheTableName, runner.req.GetKey())
+	if err != nil {
+		return nil, errors.Internal("Failed to invoke del, reason %s", err.Error())
+	}
+	return &CacheResponse{
+		Status:       database.DeletedStatus,
+		DeletedCount: deletedCount,
+	}, nil
+}
+
+func (runner *KeysQueryRunner) Run(ctx context.Context, tenant *metadata.Tenant) (*CacheResponse, error) {
+	project, err := tenant.GetDatabase(ctx, metadata.NewDatabaseNameWithBranch(runner.req.GetProject(), metadata.MainBranch))
+	if err != nil {
+		return nil, err
+	}
+	encodedCacheTableName, err := runner.encoder.EncodeCacheTableName(tenant.GetNamespace().Id(), project.Id(), runner.req.GetName())
+	if err != nil {
+		return nil, err
+	}
+
+	pattern := runner.req.GetPattern()
+	if pattern == "" {
+		pattern = "*"
+	}
+	internalKeys, err := runner.cacheStore.Keys(ctx, encodedCacheTableName, pattern)
+
+	// transform internal keys to user facing keys
+	userKeys := make([]string, len(internalKeys))
+	for index, internalKey := range internalKeys {
+		userKeys[index] = cacheKeyInternalToExternal(internalKey)
+	}
+	if err != nil {
+		return nil, errors.Internal("Failed to invoke keys, reason %s", err.Error())
+	}
+	return &CacheResponse{
+		Keys: userKeys,
+	}, nil
+}
+
+// QueryRunner is responsible for executing the current query and return the response.
+type QueryRunner interface {
+	Run(ctx context.Context, tenant *metadata.Tenant) (*CacheResponse, error)
+}
+
+// TxQueryRunner is responsible for executing the current query and return the response.
+type TxQueryRunner interface {
+	Run(ctx context.Context, tx transaction.Tx, tenant *metadata.Tenant) (*CacheResponse, context.Context, error)
+}
+
+func cacheKeyInternalToExternal(internalKey string) string {
```

Comment:
should we move this logic inside the encoder?

Reply:
good idea - moved this logic to Encoder.